File: rules/gitlab/java/xpathi/test-XpathInjection.java
// License: LGPL-3.0 License (c) find-sec-bugs
package xpathi;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;

import javax.xml.xpath.*;

public class XpathInjection {

    public static void main(String[] args) throws Exception {

        Document doc = null;

        String input = args.length != 0 ? args[1] : "guess' or '1'='1";

        String query = "//groups/group[@id='" + input + "']/writeAccess/text()";

        System.out.println(">> XPath.compile()");
        {
            XPath xpath = XPathFactory.newInstance().newXPath();
            XPathExpression expr = xpath.compile(query);
        }

        System.out.println(">> XPath.evaluate()");
        {
            XPath xpath = XPathFactory.newInstance().newXPath();
            String result = xpath.evaluate(query, doc);
            System.out.println("result=" + result);
        }

        //Safe (The next sample should not be mark)
        System.out.println(">> Safe");
        {
            XPath xpath = XPathFactory.newInstance().newXPath();
            XPathExpression expr = xpath.compile("//groups/group[@id='admin']/writeAccess/text()");
        }
    }

    public static NodeList evaluateXPath(Document doc, XPathExpression xpath) throws XPathExpressionException {
        return (NodeList) xpath.evaluate(doc, XPathConstants.NODESET);
    }

}

File: rules/gitlab/scala/xpathi/test-XpathInjection.scala
// License: LGPL-3.0 License (c) find-sec-bugs
package xpathi

import org.w3c.dom.Document
import org.w3c.dom.NodeList
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathFactory

object XpathInjection {
  @throws[Exception]
  def main(args: Array[String]): Unit = {
    val doc: Document = null
    val input = args(1)
    val query = "//groups/group[@id='" + input + "']/writeAccess/text()"
    System.out.println(">> XPath.compile()")
    XPathFactory.newInstance.newXPath.evaluate(query, doc)

    System.out.println(">> XPath.evaluate()")
    System.out.println("result=" +  XPathFactory.newInstance.newXPath.evaluate(query, doc))

    XPathFactory.newInstance.newXPath.compile(query)

    //Safe (The next sample should not be mark)
    System.out.println(">> Safe")
    XPathFactory.newInstance.newXPath.compile("//groups/group[@id='admin']/writeAccess/text()")
  }

}
